substrate-archive/src/database/queries.rs

```diff
@@ -194,6 +231,82 @@ pub(crate) fn blocks_paginated<'a>(
 	})
 }
 
+/// Get up to `max_block_load` extrinsics which are not present in the `extrinsics` table.
+/// Ordered from least to greatest number.
+pub(crate) async fn blocks_missing_extrinsics(
+	conn: &mut PgConnection,
+	max_block_load: u32,
+) -> Result<Vec<(u32, Vec<u8>, Vec<u8>, u32)>> {
+	let blocks = sqlx::query_as!(
+		BlockExtrinsics,
+		"
+		SELECT block_num, hash, ext, spec FROM blocks
+		WHERE NOT EXISTS
+			(SELECT number FROM extrinsics WHERE extrinsics.number = blocks.block_num)
```

Comment:
I wonder how this compares in performance in contrast to something like:

```
SELECT block_num, hash, ext, spec FROM blocks
LEFT OUTER JOIN extrinsics ON extrinsics.number = blocks.block_num
WHERE extrinsics.number IS NULL
```

(I 've no real reason to think that they wouldn't resolve to an identical plan; I'm just more used to using joins for this sort of thing :))

Reply:
hm, the query in archive is faster, but tbth i don't really know why, query planning in postgres is a bit of an enigma to me, i just wrote it this way because it's what i'm used to. fwiw here are the plans for both (starting with archive)

```sql
SELECT block_num, hash, ext, spec FROM blocks
WHERE NOT EXISTS
	(SELECT number FROM extrinsics WHERE extrinsics.number = blocks.block_num)
ORDER BY block_num ASC
```

```sql
"Gather Merge  (cost=1043.17..1091237.53 rows=433043 width=140) (actual time=1921.984..1984.607 rows=371882 loops=1)"
"  Workers Planned: 2"
"  Workers Launched: 2"
"  ->  Merge Anti Join  (cost=43.14..1040253.60 rows=180435 width=140) (actual time=1900.154..1932.888 rows=123961 loops=3)"
"        Merge Cond: (blocks.block_num = extrinsics.number)"
"        ->  Parallel Index Scan using blocks_block_num_index on blocks  (cost=0.43..614785.20 rows=3234289 width=140) (actual time=0.016..566.381 rows=2591749 loops=3)"
"        ->  Index Only Scan using extrinsics_number_key on extrinsics  (cost=0.43..368521.01 rows=7329251 width=4) (actual time=0.040..892.288 rows=7403366 loops=3)"
"              Heap Fetches: 4274634"
"Planning Time: 0.175 ms"
"JIT:"
"  Functions: 15"
"  Options: Inlining true, Optimization true, Expressions true, Deforming true"
"  Timing: Generation 0.990 ms, Inlining 46.963 ms, Optimization 69.031 ms, Emission 47.432 ms, Total 164.417 ms"
"Execution Time: 1993.836 ms"
``` 




```sql
EXPLAIN ANALYZE
SELECT block_num, blocks.hash, ext, spec FROM blocks
LEFT OUTER JOIN extrinsics ON extrinsics.number = blocks.block_num
WHERE extrinsics.number IS NULL
```

```sql
"Gather  (cost=376870.23..1076040.01 rows=433043 width=140) (actual time=1697.753..2332.745 rows=371882 loops=1)"
"  Workers Planned: 2"
"  Workers Launched: 2"
"  ->  Parallel Hash Anti Join  (cost=375870.23..1031735.71 rows=180435 width=140) (actual time=1676.099..2080.404 rows=123961 loops=3)"
"        Hash Cond: (blocks.block_num = extrinsics.number)"
"        ->  Parallel Seq Scan on blocks  (cost=0.00..489912.89 rows=3234289 width=140) (actual time=0.035..452.304 rows=2591749 loops=3)"
"        ->  Parallel Hash  (cost=325767.05..325767.05 rows=3053855 width=4) (actual time=698.715..698.715 rows=2467789 loops=3)"
"              Buckets: 131072  Batches: 128  Memory Usage: 3328kB"
"              ->  Parallel Index Only Scan using extrinsics_number_key on extrinsics  (cost=0.43..325767.05 rows=3053855 width=4) (actual time=0.020..404.442 rows=2467789 loops=3)"
"                    Heap Fetches: 1424878"
"Planning Time: 0.175 ms"
"JIT:"
"  Functions: 21"
"  Options: Inlining true, Optimization true, Expressions true, Deforming true"
"  Timing: Generation 1.344 ms, Inlining 46.064 ms, Optimization 39.578 ms, Emission 21.892 ms, Total 108.879 ms"
"Execution Time: 2341.940 ms"
```

the left join ends up doing a sequential scan on `blocks` which is probably why it's slower

Reply:
I've seen subqueries be faster like this before, which always struck me as a  bit magical. Joins, being "harder" to write and think through feels like they should be faster. :)

Reply:
Ooh, very interesting! It's definitely a bit of a dark art to me, too. 

I'm surprised that such different strategies were picked; I wonder whether the `ORDER BY block_num ASC` had an impact?

I do like the subquery more, so it's good to see that it's actually more performant too in this case :)